Corona-Warn-App/src/main/java/de/rki/coronawarnapp/familytest/core/repository/FamilyTestRepository.kt

```diff
@@ -1,79 +1,89 @@
 import de.rki.coronawarnapp.familytest.core.model.markBadgeAsViewed
 import de.rki.coronawarnapp.familytest.core.model.markDccCreated
 import de.rki.coronawarnapp.familytest.core.model.markViewed
-import de.rki.coronawarnapp.familytest.core.model.recycle
+import de.rki.coronawarnapp.familytest.core.model.moveToRecycleBin
 import de.rki.coronawarnapp.familytest.core.model.restore
+import de.rki.coronawarnapp.familytest.core.model.updateLabId
 import de.rki.coronawarnapp.familytest.core.model.updateResultNotification
+import de.rki.coronawarnapp.familytest.core.model.updateSampleCollectedAt
+import de.rki.coronawarnapp.familytest.core.model.updateTestResult
 import de.rki.coronawarnapp.familytest.core.storage.FamilyTestStorage
-import de.rki.coronawarnapp.tag
 import de.rki.coronawarnapp.util.TimeStamper
-import de.rki.coronawarnapp.util.coroutine.AppScope
-import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
-import de.rki.coronawarnapp.util.flow.shareLatest
-import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.first
 import kotlinx.coroutines.flow.map
-import kotlinx.coroutines.plus
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class FamilyTestRepository @Inject constructor(
-    dispatcherProvider: DispatcherProvider,
-    @AppScope private val appScope: CoroutineScope,
     private val processor: BaseCoronaTestProcessor,
     private val storage: FamilyTestStorage,
     private val timeStamper: TimeStamper,
 ) {
 
-    private val familyTestMap = storage.familyTestMap
-
     val familyTests: Flow<Set<FamilyCoronaTest>> = storage.familyTestMap.map {
-        it.values.filter { test -> test.isNotRecycled }.toSet()
-    }.shareLatest(
-        tag = TAG,
-        scope = appScope + dispatcherProvider.IO
-    )
-
-    val familyTestRecycleBin: Flow<Set<FamilyCoronaTest>> = storage.familyTestMap.map {
-        it.values.filter { test -> test.isRecycled }.toSet()
-    }.shareLatest(
-        tag = TAG,
-        scope = appScope + dispatcherProvider.IO
-    )
+        it.values.toSet()
+    }
+
+    val familyTestRecycleBin: Flow<Set<FamilyCoronaTest>> = storage.familyTestRecycleBinMap.map {
+        it.values.filter { it.isRecycled }.toSet()
+    }
 
     suspend fun registerTest(
         qrCode: CoronaTestQRCode,
         personName: String
     ): FamilyCoronaTest {
-        val test = FamilyCoronaTest(
+        return FamilyCoronaTest(
             personName = personName,
             coronaTest = processor.register(qrCode)
-        )
-        storage.save(test)
-        return test
+        ).also { test ->
+            storage.save(test)
+        }
+    }
+
+    suspend fun refresh() {
+        familyTests.first().forEach { originalTest ->
+            if (originalTest.coronaTest.isPollingStopped()) return
+
+            val updateResult = processor.pollServer(originalTest.coronaTest) ?: return
```

Comment:
what about requesting them in parallel?

Reply:
I would make the code more complicated. Do you think it is necessary for performance? I would expect that there are just very few pending tests at any moment.

Reply:
let us try first and change if necessary